### PR TITLE
pybind: build with Cython 3.0

### PR DESCRIPTION
**Pull request by kalebskeithley**
Fedora rawhide now has Cython 3.0.

N.B. rawhide has added python3-cython0.29, which can be used to pin cython to the earlier version for those who want to stay on the old version.

https://tracker.ceph.com/issues/62140

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by ljflores** [comment]
Hey @kalebskeithley, is this tracker related to (or a dupe of) https://tracker.ceph.com/issues/62291?

**Comment by kalebskeithley** [comment]
> Hey @kalebskeithley, is this tracker related to (or a dupe of) https://tracker.ceph.com/issues/62291?

That looks like a different issue.

File: src/pybind/rbd/rbd.pyx [hunk 1]
```diff
@@ -371,10 +371,10 @@ ELSE:
     cdef rados_ioctx_t convert_ioctx(rados.Ioctx ioctx) except? NULL:
         return <rados_ioctx_t>ioctx.io
 
-cdef int progress_callback(uint64_t offset, uint64_t total, void* ptr) with gil:
+cdef int progress_callback(uint64_t offset, uint64_t total, void* ptr) noexcept with gil:
     return (<object>ptr)(offset, total)
 
-cdef int no_op_progress_callback(uint64_t offset, uint64_t total, void* ptr):
+cdef int no_op_progress_callback(uint64_t offset, uint64_t total, void* ptr) noexcept with gil:
```
Comment (idryomov) [inline src/pybind/rbd/rbd.pyx:377]: What is the rationale for also adding `with gil` here?

File: src/pybind/rbd/rbd.pyx [hunk 2]
```diff
@@ -371,10 +371,10 @@ ELSE:
     cdef rados_ioctx_t convert_ioctx(rados.Ioctx ioctx) except? NULL:
         return <rados_ioctx_t>ioctx.io
 
-cdef int progress_callback(uint64_t offset, uint64_t total, void* ptr) with gil:
+cdef int progress_callback(uint64_t offset, uint64_t total, void* ptr) noexcept with gil:
```
Comment (idryomov) [inline src/pybind/rbd/rbd.pyx:374]: Reading https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html:

> ... if you are certain that your function should not raise an exception, (e.g., it does not use Python objects at all, or you plan to use it as a callback in C code that is unaware of Python exceptions), you can declare it as such using `noexcept` or by ...

`ptr` can be an arbitrary Python method/object, passed via `on_progress` parameter.  Does it qualify for the above?

Comment (kalebskeithley) [reply src/pybind/rbd/rbd.pyx:374]: By all means, suggest some alternative that will build with Cython 3. As the code stands now it does not build with Cython 3.

**Comment by kalebskeithley** [comment]
jenkins test docs

**Comment by kalebskeithley** [comment]
jenkins render docs

**Pull request closed without merging**